app/src/main/java/org/simple/clinic/scheduleappointment/ScheduleAppointmentSheetController.kt

```diff
@@ -174,4 +255,56 @@ class ScheduleAppointmentSheetController @Inject constructor(
         .filter { user -> user.loggedInStatus == User.LoggedInStatus.UNAUTHORIZED }
         .map { { ui: Ui -> ui.finish() } }
   }
+
+  private fun generatePotentialAppointmentDate(
+      appointmentDate: LocalDate,
+      potentialAppointmentDates: List<PotentialAppointmentDate>
+  ): PotentialAppointmentDate {
+    val timeToAppointment = coerceTimeToAppointmentFromPotentialsForCalendarDate(potentialAppointmentDates, appointmentDate)
+    return PotentialAppointmentDate(appointmentDate, timeToAppointment)
+  }
+
+  private fun coerceTimeToAppointmentFromPotentialsForCalendarDate(
+      potentialAppointmentDates: List<PotentialAppointmentDate>,
+      date: LocalDate
+  ): TimeToAppointment {
+    val today = LocalDate.now(clock)
+    val exactMatchingTimeToAppointment = potentialAppointmentDates
+        .find { potentialAppointmentDate -> potentialAppointmentDate.scheduledFor == date }
+        ?.timeToAppointment
+
+    return exactMatchingTimeToAppointment ?: Days(ChronoUnit.DAYS.between(today, date).toInt())
+  }
+
+  private fun generatePotentialAppointmentDate(scheduleAppointmentIn: TimeToAppointment): PotentialAppointmentDate {
+    val today = LocalDate.now(clock)
+    return PotentialAppointmentDate(today.plus(scheduleAppointmentIn), scheduleAppointmentIn)
+  }
+
+  private fun generatePotentialAppointmentDates(scheduleAppointmentsIn: List<TimeToAppointment>): List<PotentialAppointmentDate> {
+    val today = LocalDate.now(clock)
+    return scheduleAppointmentsIn
+        .map { timeToAppointment -> today.plus(timeToAppointment) to timeToAppointment }
+        .map { (appointmentDate, timeToAppointment) -> PotentialAppointmentDate(appointmentDate, timeToAppointment) }
+  }
+
+  private data class PotentialAppointmentDate(
+      val scheduledFor: LocalDate,
+      val timeToAppointment: TimeToAppointment
+  ) : Comparable<PotentialAppointmentDate> {
+    override fun compareTo(other: PotentialAppointmentDate): Int {
+      return this.scheduledFor.compareTo(other.scheduledFor)
+    }
+  }
+}
+
+private fun LocalDate.plus(timeToAppointment: TimeToAppointment): LocalDate {
```

Comment:
This could be nicer if converted to a `operator` function!

Reply:
It could, I just didn't do it because I wanted to maintain the convention of the rest of `LocalDate.plus` functions that don't have operator overloads.